### ISSUE TITLE
Cannot serde derive for generic T

### ISSUE THREAD

**Issue by e00E** [est31/serde-big-array#12]
The documentation shows how to use with crate when you have an array where the length is generic. This approach does not work for arrays where the element type is generic. This example does not compile but I would have expected it to.

```rust
#[derive(serde::Serialize, serde::Deserialize)]
struct S<T, const LEN: usize> {
    #[serde(with = "serde_big_array::BigArray")]
    array: [T; LEN],
}
```

I believe the following lines in this crate are the cause of the problem:

```rust
impl<'de, T, const N: usize> BigArray<'de> for [T; N]
    where T: Serialize + Deserialize<'de>
```

This where clause means that BigArray is only implemented if T is both Serialize and Deserialize. When serde derives Serialize and Deserialize it generates bounds for the respective trait only so BigArray is not implemented.

This it the workaround I found:

```rust
#[derive(serde::Serialize, serde::Deserialize)]
#[serde(bound(
    serialize = "for<'a> T: serde::Serialize + serde::Deserialize<'a>",
    deserialize = "T: serde::Serialize + serde::Deserialize<'de>"
))]
struct S<T, const LEN: usize> {
    #[serde(with = "serde_big_array::BigArray")]
    array: [T; LEN],
}
```

We should be able to derive Serialize and Deserialize individually even when the other one is not implemented on T. This would work if this crate added two new traits BigArraySerialize and BigArrayDeserialize which have the mentioned minimum bounds so that consumers can use `#[serde(serialize_with = "serde_big_array::BigArraySerialize")]` and `#[serde(deserialize_with = "serde_big_array::BigArrayDeserialize")]`.

**Comment by est31** [comment]
That's a good point. Splitting The trait should indeed be split up. It can be done without a breaking change, but maybe a breaking one is cleaner.

**Comment by mainrs** [comment]
I am mainly interested in a use-case where I only need to serialize data. Is there a way to do this?

**Comment by est31** [comment]
Support for separate serialization and deserialization has been implemented by #16 , which was released as `0.5.0`.
